orangecontrib/prototypes/image_grid.py

```diff
@@ -0,0 +1,148 @@
+import numpy as np
+from Orange.data import Table, Instance
+from Orange.distance import Cosine
+from Orange.preprocess import Normalize
+from Orange.projection import MDS, PCA, TSNE
+from lap import lapjv
+from scipy.spatial.distance import cdist
+from scipy.stats import kurtosis
+
+
+class ImageGrid:
+
+    def __init__(self, data):
+        self.size_x, self.size_y = 0, 0
+        self.cost, self.grid_indices, self.assignments, self.image_list = 0, [], [], []
+        self.data = data
+
+        # reduce dimensions
+        data_2dim = Table(self._reduce_dimensions(data))
+
+        # normalize the data
+        self.norm_data = np.array(self._normalize_data(data_2dim))
+
+    def process(self, size_x=0, size_y=0):
+        """
+        Process the data based on the provided grid size.
+        If the desired grid size is not provided, it is calculated from the data.
+        
+        :param size_x: The number of columns
+        :param size_y: The number of rows
+        """
+
+        # check if grid will fit all of the images
+        if size_x and size_y:
+            assert size_x * size_y >= len(self.data)
+            self.size_x, self.size_y = size_x, size_y
+
+        # calculate the necessary grid size
+        else:
+            self.size_x, self.size_y = self._get_grid_size(self.norm_data, use_default_square=False)
+
+        # calculate grid assignments and cost of the assignment
+        self.cost, self.grid_indices, self.assignments = self._get_assignments(self.norm_data)
+
+        # multiply by cell size to get [0, n) indices for grid
+        self.assignments.T[0] *= self.size_x
+        self.assignments.T[1] *= self.size_y
+
+        self.image_list = self._grid_indices_to_image_list(self.data)
+
+    @staticmethod
+    def _reduce_dimensions(data, method="MDS", use_cosine=True):
+        """
+        Reduce the dimensionality of the data to 2D.
+        
+        :param data: The image embeddings (vectors of length 2048)
+        :param method: the method to use (default MDS)
+        :param use_cosine: precompute cosine distances and pass them to MDS
+        :return: the data in 2 dimensions
+        """
+        if method is "MDS":
```

Comment:
== to compare strings.

That `"MDS" is "MDS"` is just CPython implementation detail.